File: benchmarks/tsb/bench_interval_index_query.ts
/**
 * Benchmark: IntervalIndex.indexOf / IntervalIndex.overlapping — interval lookup and overlap queries.
 * Mirrors pandas IntervalIndex.get_indexer and overlaps methods.
 * Outputs JSON: {"function": "interval_index_query", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Interval, IntervalIndex } from "../../src/index.ts";

const WARMUP = 5;
const ITERATIONS = 50;

const BREAKS = 501;
const breaks = Array.from({ length: BREAKS }, (_, i) => i * 2);
const idx = IntervalIndex.fromBreaks(breaks);

const queries = Array.from({ length: 1_000 }, (_, i) => i * 0.999);
const queryInterval = new Interval(200, 400);

for (let w = 0; w < WARMUP; w++) {
  for (const q of queries.slice(0, 50)) idx.indexOf(q);
  idx.overlapping(queryInterval);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  for (const q of queries) idx.indexOf(q);
  idx.overlapping(queryInterval);
  times.push(performance.now() - t0);
}

const total = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "interval_index_query",
    mean_ms: round3(total / ITERATIONS),
    iterations: ITERATIONS,
    total_ms: round3(total),
  }),
);

function round3(v: number): number {
  return Math.round(v * 1000) / 1000;
}

File: benchmarks/pandas/bench_interval_index_query.py
"""
Benchmark: IntervalIndex.get_indexer / IntervalIndex.overlaps — interval lookup and overlap queries.
Mirrors tsb IntervalIndex.indexOf / IntervalIndex.overlapping methods.
Outputs JSON: {"function": "interval_index_query", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd
import numpy as np

WARMUP = 5
ITERATIONS = 50

BREAKS = 501
breaks = [i * 2 for i in range(BREAKS)]
idx = pd.IntervalIndex.from_breaks(breaks)

queries = [i * 0.999 for i in range(1_000)]
query_interval = pd.Interval(200, 400)

for _ in range(WARMUP):
    idx.get_indexer(queries[:50])
    idx.overlaps(query_interval)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    idx.get_indexer(queries)
    idx.overlaps(query_interval)
    times.append((time.perf_counter() - t0) * 1000)

total = sum(times)
print(json.dumps({
    "function": "interval_index_query",
    "mean_ms": round(total / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total, 3),
}))
